connect/runtime/src/main/java/org/apache/kafka/connect/runtime/StateTracker.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime;
+
+import org.apache.kafka.connect.runtime.AbstractStatus.State;
+
+/**
+ * Utility class that tracks the current state and the duration of time spent in each state.
+ */
+public class StateTracker {
+    private volatile State currentState;
```

Comment:
A couple of questions before I dive deeper into this if I have to. 

These are a lot of volatile variables. Are we confident that the overall access pattern will be lighter than using a mutex (or combination of 1-2 volatiles with a mutex)?
Are the operations on the shared state composable? (see comment below). Are they somehow monotonically additive? (I wouldn't think so by glancing over `durationRatio` below)

Reply:
You raise a great point, and the code you reviewed definitely is not valid. I've completely redesigned the whole `StateTracker` class to be properly and efficiently threadsafe, and it should address your concerns.